### DISCUSSION TITLE
How a user code looks like (so far)

### DISCUSSION THREAD

**Discussion by fdrmrc** [fdrmrc/non_matching_test_suite#42] [Announcements]
Assume we're given a vector of agglomerations. Each agglomeration is described by a vector of cells, so `cells_to_be_agglomerated` should be a vector of vectors. They can be given all in one shot or one vector after the other. 

In the following, I'll try to describe the public interface we have so far by commenting line by line what we have in the tests.

The following three lines mark internally _masters_ and _slaves_ and create BoundingBoxes for every agglomeration.
```C++
// cells_to_be_agglomerated and cached_tria are given
FE_DGQ<2> dg_fe(1);
AgglomerationHandler<2> ah(cached_tria,dg_fe);
ah.agglomerate_cells(cells_to_be_agglomerated);
```

Then, DoFS are distributed in the triangulation on master and standard deal.II cells.

```C++
ah.initialize_hp_structure();
```

After that, for every master cell we store connectivity information about agglomerated faces. That is needed for doing `ah.reinit(cell,agglomerated_face)` in an assembly procedure.

```C++
ah.setup_connectivity_of_agglomeration();
```

This is all we need to compute the sparsity pattern associated to a Discontinuous Galerkin discretization where some cells are agglomerated together.

```C++
SparsityPattern sparsity_pattern;
ah.create_agglomeration_sparsity_pattern(sparsity);
```

**Comment by fdrmrc** [comment]
@luca-heltai This is the interface we have so far. 

Maybe we should rename `initialize_hp_structure` with `distribute_dofs_on_agglomeration`, or similar names.

Do you think that DoFs distribution (`initialize_hp_structure()`) and setup of connectivity (`setup_connectivity_of_agglomeration()`)  should be merged into one call? Something like `setup_agglomeration()`, for instance?

**Reply by luca-heltai** [reply]
I agree on simplifying. We could provide a single user entry point in the public interface, maybe `distribute_agglomerated_dofs()`, (following the convention in deal.II e.g., `distribute_mg_dofs`) that would call

```
ah.initialize_hp_structure();
dh.distribute_dofs(fe);
ah.setup_connectivity_of_agglomeration();
```

So that a user could do:
```
// cells_to_be_agglomerated and cached_tria are given
FE_DGQ<2> dg_fe(1);
AgglomerationHandler<2> ah(cached_tria);
ah.agglomerate_cells(cells_to_be_agglomerated);
ah.distribute_agglomerated_dofs(dg_fe);
```
Notice that I removed the FE from the constructor, and I moved it into the `distribute_agglomerated_dofs` argument, so that we can maybe change fe, or build it before deciding what fe to use, as in `dh(tria); dh.distribute_dofs(fe)`

**Reply by fdrmrc** [reply]
Thanks. I applied the suggestion in [555b9a6](https://github.com/fdrmrc/AggloDeal/pull/25/commits/555b9a60d4166241a709f389a7b4f684130e817b)
